File: prototypes/cel-portable/pattern.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"unicode/utf8"
)

// Same proposed ceilings as the TypeScript verifier; independent parser.
func pattern(source string) (result string, err error) {
	defer func() {
		if failure := recover(); failure != nil {
			err = fmt.Errorf("pattern: %v", failure)
		}
	}()
	chars := []rune(source)
	if !utf8.ValidString(source) || len(chars) > 4096 {
		panic("limit-or-unicode")
	}
	i := 0
	at := func() rune {
		if i >= len(chars) {
			return 0
		}
		return chars[i]
	}
	take := func() rune {
		if i >= len(chars) {
			panic("unexpected end")
		}
		c := chars[i]
		i++
		return c
	}
	if at() == '^' {
		i++
	}
	literal := func(c rune) string { return fmt.Sprintf(`\x{%x}`, c) }
	escaped := func() rune {
		c := take()
		switch c {
		case 'n':
			return '\n'
		case 'r':
			return '\r'
		case 't':
			return '\t'
		}
		if !strings.ContainsRune(`\.^$|?*+()[]{}-`, c) {
			panic("escape")
		}
		return c
	}
	type node struct {
		text         string
		cost, repeat int
	}
	var group func(int) node
	group = func(depth int) node {
		if depth > 32 {
			panic("depth")
		}
		var text strings.Builder
		cost, repeat := 0, 1
		for i < len(chars) && at() != ')' && !(at() == '$' && i == len(chars)-1) {
			atom := ""
			size, count := 1, 1
			c := take()
			if c == '|' {
				text.WriteRune(c)
				cost++
				continue
			}
			switch c {
			case '(':
				if at() == '?' && i+1 < len(chars) && chars[i+1] == ':' {
					i += 2
				}
				nested := group(depth + 1)
				if take() != ')' {
					panic("group")
				}
				atom = "(?:" + nested.text + ")"
				size = nested.cost + 1
				count = nested.repeat
			case '[':
				type item struct {
					char    rune
					escaped bool
				}
				items := []item{}
				for i < len(chars) && at() != ']' {
					isEscape := at() == '\\'
					char := take()
					if isEscape {
						char = escaped()
					}
					if char > 127 || (!isEscape && (char == '[' || char == '^' && len(items) == 0)) {
						panic("class")
					}
					items = append(items, item{char, isEscape})
				}
				if take() != ']' || len(items) == 0 {
					panic("class")
				}
				atom = "["
				for j := 0; j < len(items); j++ {
					v := items[j]
					if v.char == '-' && !v.escaped && j > 0 && j < len(items)-1 {
						panic("range")
					}
					if j+2 < len(items) && items[j+1].char == '-' && !items[j+1].escaped {
						if v.char > items[j+2].char {
							panic("range")
						}
						atom += literal(v.char) + "-" + literal(items[j+2].char)
						j += 2
					} else {
						atom += literal(v.char)
					}
				}
				atom += "]"
			case '\\':
				atom = literal(escaped())
			default:
				if strings.ContainsRune(`.^$?*+)]{}`, c) {
					panic("syntax")
				}
				atom = literal(c)
			}
			if at() != 0 && strings.ContainsRune("?*+", at()) {
				atom += string(take())
				size++
			} else if at() == '{' {
				i++
				lower, upper := "", ""
				for at() >= '0' && at() <= '9' {
					lower += string(take())
				}
				comma := false
				if at() == ',' {
					comma = true
					i++
					for at() >= '0' && at() <= '9' {
						upper += string(take())
					}
				}
				lo, e := strconv.Atoi(lower)
				hi, he := strconv.Atoi(upper)
				if lower == "" || take() != '}' || e != nil || lo > 1000 || (upper != "" && (he != nil || hi > 1000 || lo > hi)) {
					panic("repeat")
				}
				multiplier, nesting := max(1, lo), max(1, lo)
				if comma {
					if upper != "" {
						multiplier = max(1, hi)
						nesting = max(1, hi)
					} else {
						multiplier = max(1, lo+1)
					}
				}
				count *= nesting
				size *= multiplier
				atom += "{" + strconv.Itoa(lo)
				if comma {
					atom += ","
					if upper != "" {
						atom += strconv.Itoa(hi)
					}
				}
				atom += "}"
			}
			cost += size
			repeat = max(repeat, count)
			if cost > 8192 || repeat > 1000 {
				panic("expansion")
			}
			text.WriteString(atom)
		}
		return node{text.String(), cost, repeat}
	}
	compiled := group(0)
	if at() == '$' {
		i++
	}
	if i != len(chars) {
		panic("syntax")
	}
	return `\A(?:` + compiled.text + `)\z`, nil
}

File: prototypes/cel-portable/pattern.ts
// PROTOTYPE: proposed operational limits, not new domain constraints.
export function pattern(source: string): string {
  if (!source.isWellFormed() || [...source].length > 4096) throw Error('pattern-limit-or-unicode');
  const chars = [...source];
  let i = 0;
  if (chars[0] === '^') i++;
  function literal(c: string) { return '\\x{' + c.codePointAt(0)!.toString(16) + '}'; }
  function escaped(): string {
    const c = chars[i++];
    if (c === 'n') return '\n';
    if (c === 'r') return '\r';
    if (c === 't') return '\t';
    if (!c || !'\\.^$|?*+()[]{}-'.includes(c)) throw Error('pattern-escape');
    return c;
  }
  function group(depth: number): {text: string, cost: number, repeat: number} {
    if (depth > 32) throw Error('pattern-depth');
    let text = '', cost = 0, repeat = 1;
    while (i < chars.length && chars[i] !== ')' && !(chars[i] === '$' && i === chars.length - 1)) {
      let atom = '', size = 1, count = 1;
      const c = chars[i++];
      if (c === '|') { text += '|'; cost++; continue; }
      if (c === '(') {
        if (chars[i] === '?' && chars[i + 1] === ':') i += 2;
        const nested = group(depth + 1);
        if (chars[i++] !== ')') throw Error('pattern-group');
        atom = '(?:' + nested.text + ')'; size = nested.cost + 1; count = nested.repeat;
      } else if (c === '[') {
        const items: {char: string, escaped: boolean}[] = [];
        while (i < chars.length && chars[i] !== ']') {
          const escapedChar = chars[i] === '\\';
          let char = chars[i++];
          if (escapedChar) char = escaped();
          if (char.codePointAt(0)! > 127 || (!escapedChar && (char === '[' || char === '^' && !items.length))) throw Error('pattern-class');
          items.push({char, escaped: escapedChar});
        }
        if (chars[i++] !== ']' || !items.length) throw Error('pattern-class');
        atom = '[';
        for (let j = 0; j < items.length; j++) {
          const item = items[j];
          if (item.char === '-' && !item.escaped && j > 0 && j < items.length - 1) throw Error('pattern-range');
          if (j + 2 < items.length && items[j + 1].char === '-' && !items[j + 1].escaped) {
            if (item.char > items[j + 2].char) throw Error('pattern-range');
            atom += literal(item.char) + '-' + literal(items[j + 2].char); j += 2;
          } else atom += literal(item.char);
        }
        atom += ']';
      } else if (c === '\\') atom = literal(escaped());
      else {
        if ('.^$?*+)]{}'.includes(c)) throw Error('pattern-syntax');
        atom = literal(c);
      }
      if (['?', '*', '+'].includes(chars[i])) { atom += chars[i++]; size++; }
      else if (chars[i] === '{') {
        i++;
        let lower = '', upper = '';
        while (/[0-9]/.test(chars[i] ?? '')) lower += chars[i++];
        let comma = false;
        if (chars[i] === ',') { comma = true; i++; while (/[0-9]/.test(chars[i] ?? '')) upper += chars[i++]; }
        if (!lower || chars[i++] !== '}' || +lower > 1000 || (upper && (+upper > 1000 || +lower > +upper))) throw Error('pattern-repeat');
        const multiplier = Math.max(1, comma ? (upper ? +upper : +lower + 1) : +lower);
        count *= Math.max(1, comma ? (upper ? +upper : +lower) : +lower);
        size *= multiplier;
        atom += '{' + Number(lower) + (comma ? ',' + (upper ? Number(upper) : '') : '') + '}';
      }
      cost += size; repeat = Math.max(repeat, count);
      if (cost > 8192 || repeat > 1000) throw Error('pattern-expansion');
      text += atom;
    }
    return {text, cost, repeat};
  }
  const result = group(0);
  if (chars[i] === '$') i++;
  if (i !== chars.length) throw Error('pattern-syntax');
  return '\\A(?:' + result.text + ')\\z';
}
